core/example/03_struct_of_arrays/soa_example.cpp

```diff
@@ -0,0 +1,142 @@
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Cabana_Core.hpp>
+
+#include <iostream>
+
+//---------------------------------------------------------------------------//
+// SoA example.
+//---------------------------------------------------------------------------//
+void soaExample()
+{
+    /*
+      Cabana structs-of-arrays (SoAs) are groups of Tuples with the data
+      specified in an order such that the tuple index is stride-1 in
+      memory. This results in each dimension of each multidimensional tuple
+      member being stored in a contiguous memory block the size of the number
+      of tuples. The number of tuples in an SoA is defined as the "vector
+      length".
+
+      For example, consider declaring the following member type to store two
+      arrays of doubles in each tuple:
+
+          using types = Cabana::MemberTypes<double[3][2],double[4]>;
+
+      and vector length:
+
+          const int veclen = 8;
+
+      Declaring a Cabana SoA as Cabana::SoA<types,veclen> gives the equivalent
+      data layout:
+
+          struct MyEquivalentSoA
+          {
+              double d0[2][3][veclen];
+              double d1[4][veclen];
+          };
+
+      Note: The data in this struct definition with an equivalent memory
+      layout is stride-1 in the vector index.
+
+      Note: The higher dimensional indices have their order reversed (see the
+      order of the data of member `d0` in the equivalent struct
+      definition). This is to create a Fortran-equivalent or column-major
+      ordering of the data relative to the vector index.
+
+      Note: When defined as a compile-time argument the vector length must be
+      a power of 2. Vector lengths that are not a power of 2 will emit a
+      compiler error.
+    */
+
+    /* Start by declaring the types in our tuples will store. Store a rank-2
+       array of doubles, a rank-1 array of floats, and a single integer in
+       each tuple.
+    */
+    using DataTypes = Cabana::MemberTypes<double[3][3],
+                                          float[4],
+                                          int>;
+
+    /*
+      Next declare the vector length of our SoA. This is how many tuples the
+      SoA will contain. A reasonable number for performance should be some
+      multiple of the vector length on the machine you are using.
+    */
+    const int VectorLength = 4;
+
```

Comment:
As a side thought: I am wondering if there is a way to automatically figure out the vector length supported by the device where the code compiles? If there is, then one can just say 
const int VectorLength = DefaultVeclen

Reply:
So we have presets for this - the user doesn't have to define it. In the future we can automatically select the vector length based on the device type flags we get from kokkos. Should be pretty easy to do. I will make an issue for this.